lib/AST/RequirementMachine/RequirementMachineRequests.cpp

```diff
@@ -850,9 +856,33 @@ InferredGenericSignatureRequestRQM::evaluate(
                                 allowConcreteGenericParams
                                 ? AllowConcreteTypePolicy::All
                                 : AllowConcreteTypePolicy::AssocTypes);
-    }
 
-    // FIXME: Handle allowConcreteGenericParams
+      if (!allowConcreteGenericParams) {
```

Comment:
The FIXME was in the wrong place :) This should happen after the installRequirementMachine() call below, otherwise we'll always create a new requirement machine for this signature even though we have one on-hand here.

Alternatively, use the RequirementMachine::getCanonicalTypeInContext() method instead of calling it on the signature.